misc/vtk/src/processors/vtktovolume.cpp

```diff
@@ -0,0 +1,422 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <inviwo/vtk/processors/vtktovolume.h>
+#include <inviwo/core/util/filesystem.h>
+#include <inviwo/core/datastructures/volume/volume.h>
+#include <inviwo/core/datastructures/volume/volumeramprecision.h>
+#include <inviwo/core/common/inviwoapplication.h>
+#include <inviwo/core/network/networklock.h>
+#include <inviwo/core/util/clock.h>
+#include <inviwo/core/util/formats.h>
+#include <modules/base/algorithm/dataminmax.h>
+
+#include <warn/push>
+#include <warn/ignore/all>
+#include <vtkCallbackCommand.h>
+#include <warn/pop>
+
+#include <vtkInformation.h>
+#include <vtkDataSet.h>
+#include <vtkCellData.h>
+#include <vtkPointData.h>
+#include <vtkDataArray.h>
+#include <vtkRectilinearGrid.h>
+#include <vtkStructuredGrid.h>
+#include <vtkImageData.h>
+#include <vtkMatrix4x4.h>
+#include <vtkType.h>
+
+#include <fstream>
+#include <inviwo/vtk/util/vtkutil.h>
+#include <fmt/format.h>
+
+namespace inviwo {
+
+// The Class Identifier has to be globally unique. Use a reverse DNS naming scheme
+const ProcessorInfo VTKtoVolume::processorInfo_{
+    "org.inviwo.VTKtoVolume",     // Class identifier
+    "VTK to Volume",              // Display name
+    "VTK",                        // Category
+    CodeState::Experimental,      // Code state
+    "CPU, Data Conversion, VTK",  // Tags
+};
+const ProcessorInfo VTKtoVolume::getProcessorInfo() const { return processorInfo_; }
+
+VTKtoVolume::VTKtoVolume()
+    : Processor()
+    , outport_("VolumeOutport")
+    , inport_("VTKDataObjectInport")
+    , useCellData_("useCellData", "Use cell data", false)
+    , dataArrays_("availableArrays", "Data Arrays")
+    , dataFormats_("existingDataFormats", "Data formats")
+    , convertButton_("triggerConvert", "Convert")
+    , dataRange_("dataRange", "Data range", vec2(0, 1))
+    , reactToChange_(true)
+    , convertData_(false)
+    , formerArraySelection_() {
+
+    addPort(outport_);
+    addPort(inport_);
+    addProperty(useCellData_);
+    addProperty(dataFormats_);
+    addProperty(dataArrays_);
+    addProperty(convertButton_);
+    addProperty(dataRange_);
+
+    inport_.onChange([this]() {
+        if (reactToChange_) updateFormats();
```

Comment:
This feels quite error prone, i.e. one would forget this and wonder why the output is not up-to-date. Is it really needed?

Reply:
The general problem is that we can only know the data types and vtkDataArray names at runtime, when we have data.
This information is necessary since we need to pick arrays to include. Thus, in the user interface:
* Choose to use point or cell data.
* Pick a DataFormat, since inviwo volumes only support one format at a time.
  * This updates the data array names. Without the dirty flag `reactToChange_`, every addition would trigger `onChange()`
* If there are more than 4 arrays, pick a subset, since volumes only support up to 4 channels.
  * If 4 channels are selected, the other data arrays are marked as readonly, so no 5th channel can be chosen.

Now, since the conversion is quite expensive, we don't want to trigger it whenever a data array is added or removed. Thus there is a button for the `convertData()` function. Another possibility is to have a `lock` or conversely `auto-update` switch, but the button is more in line with what we have for some loaders. 
`convertData()` is also called when a previous setting is applied, either from deserialization or changing to an input which includes all data arrays that were previously selected (e.g., the next step in a time series).

Reply:
Do I understand it correctly if I say that the reactToChange exist to prevent updating when editing this processor's properties?


> the button is more in line with what we have for some loaders.

True, but for the loaders they always read data the first time so the button is more for manual reloading. The button in this processor would need to be pressed in order for the output to be valid? The suggested auto-update is probably more intuitive in that case.

Reply:
Yes, the flag does exactly that.

> True, but for the loaders they always read data the first time so the button is more for manual reloading.

If we load settings (i.e., deserialize or changed input being very similar to previous), one conversion is triggered.
If it's completely new data, I would not want to trigger a conversion. In general, there are several formats to be chosen from. And for more complex data sets, more channels than can be stored in an inviwo volume.
My example data has 2 'spurious' arrays in uint8 and int8, and 11 float arrays. It's hard to make a guess what the user would want to convert in that case, and just converting _something_ is a huge waste of computation time.
Turning on a potential `auto-update` flag from the beginning would also mean that every small action by the user triggers such an update. Thus, I know that I would turn on `auto-update` to trigger exactly one change, and then turn it off immediately to not accidentally trigger a convert.

Can you tell me what you optimal workflow for this would be, if there was an `auto-update` switch? Would it replace the button, or would they co-exist?

Reply:
> Can you tell me what you optimal workflow for this would be, if there was an auto-update switch? Would it replace the button, or would they co-exist?

I see. If the auto-update would be off by default it would mean that you would have kind of the same behaviour? I.e. only having an auto-update but have it off by default.
 You know the use case better. I just from experience know that having these out-of-state thingies can cause a lot of headache because one forgot to 'press this button'

Reply:
> I see. If the auto-update would be off by default it would mean that you would have kind of the same behaviour?

Yes.
I can move the button to the top, making it more obvious.

Reply:
This is the view now. The text changes if at least one array is selected.
![Screenshot from 2020-03-18 12-06-42](https://user-images.githubusercontent.com/2829051/76954763-3010f680-6911-11ea-8d49-4017319f8473.png)

Reply:
the values (1) refers to the number of components/channels?
Anyway, looks good now I think.

Reply:
Yes, it's the number of channels. No idea how to make that more obvious without having very long strings.